File: src/lib/ipc/communication.ts
import { invoke } from '@tauri-apps/api/core';

export interface MllpSendResult {
	success: boolean;
	response: string;
	response_time_ms: number;
	error: string | null;
}

export interface MllpReceivedMessage {
	content: string;
	source_addr: string;
	received_at: string;
}

export interface HttpResult {
	success: boolean;
	status_code: number;
	status_text: string;
	headers: Record<string, string>;
	body: string;
	response_time_ms: number;
	error: string | null;
}

export interface ConnectionProfile {
	id: string;
	name: string;
	profile_type: 'mllp' | 'http';
	host: string;
	port: number;
	timeout_secs: number;
	url: string | null;
	headers: string | null;
	auto_ack: boolean;
}

export interface HistoryEntry {
	id: string;
	profile_name: string;
	profile_type: string;
	direction: string;
	content_preview: string;
	status: string;
	response_time_ms: number;
	timestamp: string;
}

// --- MLLP ---
export interface MllpSendOptions {
	/** TCP connect timeout (seconds). */
	timeoutSecs?: number;
	/** ACK read timeout (seconds); defaults to timeoutSecs backend-side. */
	responseTimeoutSecs?: number;
	encoding?: string;
	/** Framing byte overrides as hex strings ("0x0B"); invalid values fall back to standard MLLP. */
	startChar?: string;
	endChar1?: string;
	endChar2?: string;
	profileName?: string;
}

export async function mllpSend(
	host: string, port: number, message: string,
	opts: MllpSendOptions = {},
): Promise<MllpSendResult> {
	return invoke('mllp_send', {
		host, port, message,
		timeoutSecs: opts.timeoutSecs,
		responseTimeoutSecs: opts.responseTimeoutSecs,
		encoding: opts.encoding,
		startChar: opts.startChar,
		endChar1: opts.endChar1,
		endChar2: opts.endChar2,
		profileName: opts.profileName,
	});
}

export async function mllpReceive(
	port: number, timeoutSecs?: number, autoAck?: boolean,
): Promise<MllpReceivedMessage> {
	return invoke('mllp_receive', { port, timeoutSecs, autoAck });
}

// --- Persistent MLLP listener ---
export interface ListenerConfig {
	port: number;
	bind_address: string;
	auto_ack: boolean;
	ack_code: string;       // 'AA' | 'AE' | 'AR'
	read_timeout_secs: number;
	encoding: string;       // 'UTF-8' | 'ISO-8859-1' | 'windows-1252' | etc.
}

export interface ListenerStatus {
	running: boolean;
	port: number | null;
	bind_address: string | null;
}

/** Payload of the `mllp:received` Tauri event. */
export interface MllpReceivedEvent {
	content: string;
	source_addr: string;
	received_at: string;
	/** Payload size after MLLP unframing, in bytes. */
	bytes: number;
	/** ACK code sent back ("AA"/"AE"/"AR"), or null when auto-ACK is off. */
	ack_code: string | null;
	/** Charset the payload was decoded with. */
	encoding: string;
}

export async function mllpListenStart(config: ListenerConfig): Promise<ListenerStatus> {
	return invoke('mllp_listen_start', { config });
}

export async function mllpListenStop(): Promise<ListenerStatus> {
	return invoke('mllp_listen_stop');
}

export async function mllpListenStatus(): Promise<ListenerStatus> {
	return invoke('mllp_listen_status');
}

// --- HTTP ---
export async function httpRequest(
	url: string, method: string,
	headers?: Record<string, string>, body?: string,
	timeoutSecs?: number, followRedirects?: boolean, profileName?: string,
): Promise<HttpResult> {
	return invoke('http_request', { url, method, headers, body, timeoutSecs, followRedirects, profileName });
}

// --- ACK ---
export async function generateAck(
	ackCode: string, messageControlId: string, textMessage?: string,
): Promise<string> {
	return invoke('generate_ack', { ackCode, messageControlId, textMessage });
}

// --- Profiles ---
export async function saveConnectionProfile(profile: ConnectionProfile): Promise<void> {
	return invoke('save_connection_profile', { profile });
}

export async function getConnectionProfiles(): Promise<ConnectionProfile[]> {
	return invoke('get_connection_profiles');
}

export async function deleteConnectionProfile(id: string): Promise<void> {
	return invoke('delete_connection_profile', { id });
}

// --- History ---
export async function getRequestHistory(limit?: number): Promise<HistoryEntry[]> {
	return invoke('get_request_history', { limit });
}

export async function clearRequestHistory(): Promise<void> {
	return invoke('clear_request_history');
}

File: src-tauri/src/commands/communication.rs
use std::collections::HashMap;
use tauri::State;

use crate::communication::http_client::{self, HttpMethod, HttpResult};
use crate::communication::mllp::{self, MllpSendResult};
use crate::communication::mllp_listener::{ListenerConfig, ListenerState, ListenerStatus};
use crate::communication::profiles::{ConnectionProfile, HistoryEntry};
use crate::database::Database;
use crate::licensing::feature_gate;

// --- MLLP Commands ---

/// Parse a framing-byte override like "0x0B" / "0B" (hex). Anything
/// unparsable falls back to the standard MLLP byte, so a typo in the
/// advanced settings can't silently produce unframeable traffic.
fn parse_framing_byte(input: &Option<String>, default: u8) -> u8 {
    input
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .and_then(|s| {
            let hex = s.strip_prefix("0x").or_else(|| s.strip_prefix("0X")).unwrap_or(s);
            u8::from_str_radix(hex, 16).ok()
        })
        .unwrap_or(default)
}

#[tauri::command]
pub async fn mllp_send(
    host: String,
    port: u16,
    message: String,
    timeout_secs: Option<u64>,
    response_timeout_secs: Option<u64>,
    encoding: Option<String>,
    start_char: Option<String>,
    end_char1: Option<String>,
    end_char2: Option<String>,
    profile_name: Option<String>,
    db: State<'_, Database>,
) -> Result<MllpSendResult, String> {
    feature_gate::require("mllp_send")?;

    let connect_timeout = timeout_secs.unwrap_or(30);
    let opts = mllp::SendOptions {
        connect_timeout_secs: connect_timeout,
        response_timeout_secs: response_timeout_secs.unwrap_or(connect_timeout),
        encoding: encoding.unwrap_or_default(),
        start_byte: parse_framing_byte(&start_char, mllp::MLLP_START),
        end_byte_1: parse_framing_byte(&end_char1, mllp::MLLP_END_1),
        end_byte_2: parse_framing_byte(&end_char2, mllp::MLLP_END_2),
    };
    let result = mllp::send_with_options(&host, port, &message, &opts).await;

    let preview: String = message.chars().take(100).collect();
    let status = if result.success { "OK" } else { "FAILED" };
    let entry = HistoryEntry {
        id: uuid::Uuid::new_v4().to_string(),
        profile_name: profile_name.unwrap_or_else(|| format!("{}:{}", host, port)),
        profile_type: "mllp".into(),
        direction: "send".into(),
        content_preview: preview,
        status: status.into(),
        response_time_ms: result.response_time_ms,
        timestamp: chrono::Utc::now().to_rfc3339(),
    };
    let _ = db.add_history_entry(&entry);

    Ok(result)
}

// --- Persistent MLLP listener (start / stop / status) -------------------------
//
// The listener keeps accepting connections and emits a Tauri event
// (`mllp:received`) for each incoming message until the user calls stop. This
// is what the Communication panel's Listen / Stop button drives.
// (The old single-shot `mllp_receive` IPC command was removed in 0.7.0 — it
// was never reachable from the UI once the persistent listener shipped. The
// underlying `mllp::receive_one` stays as a tested library primitive.)

#[tauri::command]
pub async fn mllp_listen_start(
    app: tauri::AppHandle,
    state: tauri::State<'_, ListenerState>,
    config: ListenerConfig,
) -> Result<ListenerStatus, String> {
    feature_gate::require("mllp_listen")?;
    state.start(app, config).await
}

#[tauri::command]
pub async fn mllp_listen_stop(
    state: tauri::State<'_, ListenerState>,
) -> Result<ListenerStatus, String> {
    Ok(state.stop().await)
}

#[tauri::command]
pub async fn mllp_listen_status(
    state: tauri::State<'_, ListenerState>,
) -> Result<ListenerStatus, String> {
    Ok(state.status().await)
}

// --- HTTP Commands ---

#[tauri::command]
pub async fn http_request(
    url: String,
    method: String,
    headers: Option<HashMap<String, String>>,
    body: Option<String>,
    timeout_secs: Option<u64>,
    follow_redirects: Option<bool>,
    profile_name: Option<String>,
    db: State<'_, Database>,
) -> Result<HttpResult, String> {
    let http_method = HttpMethod::from_str(&method)
        .ok_or_else(|| format!("Invalid HTTP method: {}", method))?;

    // GET is community; POST/PUT/DELETE/PATCH require Pro
    match http_method {
        HttpMethod::Get => feature_gate::require("http_get")?,
        _ => feature_gate::require("http_mutate")?,
    }

    // Auth headers require Pro
    let hdrs = headers.unwrap_or_default();
    if hdrs.keys().any(|k| k.to_lowercase() == "authorization") {
        feature_gate::require("http_auth")?;
    }

    let timeout = timeout_secs.unwrap_or(30);
    let result = http_client::send_request(
        &url,
        http_method,
        &hdrs,
        body.as_deref(),
        timeout,
        follow_redirects.unwrap_or(true),
    ).await;

    let preview: String = body.as_deref().unwrap_or("").chars().take(100).collect();
    let status = if result.success {
        format!("{} {}", result.status_code, result.status_text)
    } else {
        "FAILED".into()
    };
    let entry = HistoryEntry {
        id: uuid::Uuid::new_v4().to_string(),
        profile_name: profile_name.unwrap_or_else(|| url.clone()),
        profile_type: "http".into(),
        direction: "send".into(),
        content_preview: format!("{} {} | {}", method.to_uppercase(), url, preview),
        status,
        response_time_ms: result.response_time_ms,
        timestamp: chrono::Utc::now().to_rfc3339(),
    };
    let _ = db.add_history_entry(&entry);

    Ok(result)
}

// --- ACK Generation ---

#[tauri::command]
pub fn generate_ack(
    ack_code: String,
    message_control_id: String,
    text_message: Option<String>,
) -> Result<String, String> {
    Ok(crate::parser::hl7::ack::generate_ack(
        &ack_code,
        &message_control_id,
        "BridgeLab",
        "RemoteApp",
        text_message.as_deref(),
    ))
}

// --- Connection Profiles ---

#[tauri::command]
pub fn save_connection_profile(
    profile: ConnectionProfile,
    db: State<'_, Database>,
) -> Result<(), String> {
    db.save_connection_profile(&profile)
}

#[tauri::command]
pub fn get_connection_profiles(db: State<'_, Database>) -> Result<Vec<ConnectionProfile>, String> {
    db.get_connection_profiles()
}

#[tauri::command]
pub fn delete_connection_profile(id: String, db: State<'_, Database>) -> Result<(), String> {
    db.delete_connection_profile(&id)
}

#[tauri::command]
pub fn get_request_history(
    limit: Option<usize>,
    db: State<'_, Database>,
) -> Result<Vec<HistoryEntry>, String> {
    db.get_request_history(limit.unwrap_or(50))
}

#[tauri::command]
pub fn clear_request_history(db: State<'_, Database>) -> Result<(), String> {
    db.clear_request_history()
}
